File: testSuite/test-concentration-Ludlow2016.pl
#!/usr/bin/env perl
use strict;
use warnings;
use lib $ENV{'GALACTICUS_EXEC_PATH'}."/perl";
use PDL;
use PDL::NiceSlice;
use PDL::IO::HDF5;
use PDL::Stats::Basic;
use System::Redirect;
use File::Slurp qw(slurp);

# Run models that check that the results of Benson, Ludlow, & Cole (2019; http://adsabs.harvard.edu/abs/2019MNRAS.485.5010B)
# applying the Ludlow et al. (2016; http://adsabs.harvard.edu/abs/2016MNRAS.460.1214L) concentration model to merger trees can be
# reproduced.
# Andrew Benson (27-July-2021)

# Make output directory.
system("mkdir -p outputs/");

# Specify the tests to run. Mean and tolerance targets are taken from Table 2 of Benson, Ludlow, & Cole (2019).
my @tests =
    (
     {
	 # Model without environmental dependence.
	 suffix           => "",
	 mean             => 1.104,
	 meanTolerance    => 0.040,
	 scatter          => 0.103,
	 scatterTolerance => 0.005
     },
     {
	 # Model with environmental dependence.
	 suffix           => "environmental",
	 mean             => 1.077,
	 meanTolerance    => 0.040,
	 scatter          => 0.114,
	 scatterTolerance => 0.005
     }
    );

# Iterate over tests.
foreach my $test ( @tests ) {

    # Run the model.
    &System::Redirect::tofile("cd ..; Galacticus.exe testSuite/parameters/concentrationDistributionLudlow2016".ucfirst($test->{'suffix'}).".xml","outputs/concentrationDistributionLudlow2016".ucfirst($test->{'suffix'}).".log");
    unless ( $? == 0 ) {
    	print "FAILED: Ludlow2016 ".$test->{'suffix'}." concentration model failed to run\n";
    	system("cat outputs/concentrationDistributionLudlow2016".ucfirst($test->{'suffix'}).".log");
    	exit;
    }
    
    # Select halos in the mass range used for the results in Table 2 of Benson, Ludlow, & Cole (2019), and compute the mean and
    # scatter in concentration.
    my $model          = new PDL::IO::HDF5("outputs/concentrationDistributionLudlow2016".ucfirst($test->{'suffix'}).".hdf5");
    my $outputs        = $model   ->group  ('Outputs'                )       ;
    my $output         = $outputs ->group  ('Output1'                )       ;
    my $nodeData       = $output  ->group  ('nodeData'               )       ;
    my $nodeIsIsolated = $nodeData->dataset('nodeIsIsolated'         )->get();
    my $massHalo       = $nodeData->dataset('massHaloEnclosedCurrent')->get();
    my $concentration  = $nodeData->dataset('concentration'          )->get();
    # Compute N-body measurement uncertainties to both halo masses and concentrations.
    my $massParticle             = pdl 1.6e5; # COCO simulation particle mass.
    my $countParticles           = $massHalo/$massParticle;
    my $alpha                    = -0.20+1.46*log10($concentration)-0.25*log10($concentration)**2;
    my $b                        = -0.54;
    my $concentrationUncertainty = 10.0**($alpha+$b*log10($countParticles)); # Concentration uncertainty model from equation 5 of Benson, Ludlow, & Cole (2019).
    my $massUncertainty          = 0.135*(1000.0/$countParticles)**(1.0/3.0); # Mass uncertainty model from Trenti et al. (2010; http://adsabs.harvard.edu/abs/2012ApJ...747..100S).
    # Construct perturbed halo masses.
    my $massMeasured = $massHalo*exp($massUncertainty*grandom(nelem($massHalo)));
    # Select halos for inclusion.
    my $selection = which(($nodeIsIsolated == 1) & ($massMeasured->log10() >= 9.402) & ($massMeasured->log10() < 9.902));
    # Construct perturbed concentrations.
    my $concentrationsMeasured = $concentration->log10()->($selection)+$concentrationUncertainty->($selection)*grandom(nelem($selection));
    # Evaluate the mean and scatter and check these match the target values.
    my $mean    = $concentrationsMeasured->average();
    my $scatter = $concentrationsMeasured->stdv();
    my $status  =
	(
	 abs($mean   -$test->{'mean'   }) < $test->{'meanTolerance'   }
	 &&
	 abs($scatter-$test->{'scatter'}) < $test->{'scatterTolerance'}
	)
	? "SUCCESS"
	: "FAILED";
    print $status.": Ludlow2016 ".$test->{'suffix'}." concentration\n";
}

exit;

File: testSuite/test-concentration-Ludlow2016.py
#!/usr/bin/env python3
import subprocess
import sys
import h5py
import numpy as np

# Run models that check that the results of Benson, Ludlow, & Cole (2019) can be reproduced.
# Andrew Benson (ported to Python)

np.random.seed(42)

# Make output directory.
subprocess.run("mkdir -p outputs/", shell=True)

# Specify the tests to run. Mean and tolerance targets are taken from Table 2 of Benson, Ludlow, & Cole (2019).
tests = [
    {"suffix": "",              "mean": 1.104, "meanTolerance": 0.040, "scatter": 0.103, "scatterTolerance": 0.008},
    {"suffix": "environmental", "mean": 1.077, "meanTolerance": 0.040, "scatter": 0.114, "scatterTolerance": 0.016},
]

for test in tests:
    suffix        = test["suffix"]
    suffixCapital = suffix[0].upper() + suffix[1:] if suffix else ""

    # Run the model.
    status = subprocess.run(
        f"cd ..; ./Galacticus.exe testSuite/parameters/concentrationDistributionLudlow2016{suffixCapital}.xml",
        shell=True
    )
    if status.returncode != 0:
        print(f"FAILED: Ludlow2016 {suffix} concentration model failed to run")
        sys.exit(0)

    # Select halos in the mass range used for the results in Table 2 of Benson, Ludlow, & Cole (2019).
    with h5py.File(f"outputs/concentrationDistributionLudlow2016{suffixCapital}.hdf5", "r") as model:
        nodeData      = model["Outputs/Output1/nodeData"]
        nodeIsIsolated= nodeData["nodeIsIsolated"][:]
        massHalo      = nodeData["massHaloEnclosedCurrent"][:]
        concentration = nodeData["concentration"][:]

    # Compute N-body measurement uncertainties.
    massParticle             = 1.6e5  # COCO simulation particle mass.
    countParticles           = massHalo / massParticle
    alpha                    = -0.20 + 1.46 * np.log10(concentration) - 0.25 * np.log10(concentration)**2
    b                        = -0.54
    concentrationUncertainty = 10.0**(alpha + b * np.log10(countParticles))
    massUncertainty          = 0.135 * (1000.0 / countParticles)**(1.0/3.0)

    # Construct perturbed halo masses.
    massMeasured = massHalo * np.exp(massUncertainty * np.random.randn(len(massHalo)))

    # Select halos for inclusion.
    selection = np.where(
        (nodeIsIsolated == 1)
        & (np.log10(massMeasured) >= 9.402)
        & (np.log10(massMeasured)  < 9.902)
    )[0]

    # Construct perturbed concentrations.
    concentrationsMeasured = np.log10(concentration[selection]) + concentrationUncertainty[selection] * np.random.randn(len(selection))

    # Evaluate the mean and scatter.
    mean    = np.mean(concentrationsMeasured)
    scatter = np.std(concentrationsMeasured)
    print(f"Mean - {mean} (target = {test['mean']} \u00b1 {test['meanTolerance']})")
    print(f"Scatter - {scatter} (target = {test['scatter']} \u00b1 {test['scatterTolerance']})")
    status_str = "SUCCESS" if (
        abs(mean    - test["mean"])    < test["meanTolerance"] and
        abs(scatter - test["scatter"]) < test["scatterTolerance"]
    ) else "FAILED"
    print(f"{status_str}: Ludlow2016 {suffix} concentration")
